### PR TITLE
Implement Normal, a solver applying an inner solver to the normal equations

### PR DESCRIPTION
**Pull request by adconner**
This updates #159 to be independent of #158, as requested there.

### CODE REVIEW

**Comment by adconner** [comment]
One question about the implementation of `allow_dependent_{rows,columns}`: Currently I implement them defensively with regard to the possibility that the solver state might be transposed, taking into account the possibility that the user manually constructs the solver state and transposes it (for instance, maybe they want to efficiently apply a linear solve to many systems, some involving A and some involving A^t). However, thinking more closely about the way the library works, I think they only way they can do this is to directly call `compute` on the solver state object, and in this usage, they do not get the custom jvp (any gradients computed would be directly differentiating the compute call), so the result of `allow_dependent_{rows,columns}` is irrelevant. Any calls to the `allow_dependent_{rows,columns}` in the library itself follow from the jvp of a `linear_solve` call, which seems to mean in these functions the solver state can be assumed freshly constructed from the given operator. 

So is it correct to say that the result of allow_dependent_{rows,columns} can assume the eventually used solver state is fresh, ie unmodified by transpose or conj? In this case, this assumption would allow a more precise and simpler implementation, removing the special case for square operators. If you agree I can make the simplification.

File: lineax/_solve.py [hunk 1]
```diff
@@ -400,6 +400,9 @@ def allow_dependent_columns(self, operator: AbstractLinearOperator) -> bool:
         If `True` then a more expensive backward pass is needed, to account for the
         extra generality.
 
+        The value `True` does not guarantee that the solver will produce correct
```
[the remark this answers is not in the document]
Comment (adconner) [reply lineax/_solve.py:403]: I think it cant be as you say exactly. Consider unpivoted qr for a tall matrix. We have `allow_dependent_rows = True` and `allow_dependent_cols = False` and the method supports full rank matrices only. If the input matrix is not full rank, it may give incorrect (not NaN) pseudoinverse results. 

But I hear you that there is intended to be some positive assertion for the `True` value. Is it that if they are both true then pseudoinverses must be fully supported?

[the remark this answers is not in the document]
Comment (adconner) [reply lineax/_solve.py]: Its enough for instance to use this structurally non-full-rank example (I've modified the lineax tests in this change to use this example as well). For a fully generic nonstructural rank 2, uncomment the alternative lines.

```python
import lineax as lx
import jax.numpy as jnp
import jax.random as jr

want_full_rank = False

key1, key2 = jr.split(jr.key(567), 2)
x = jr.normal(key1, (5, 3))
y = jr.normal(key2, (5,))

if not want_full_rank:
    x = x.at[1:,1:].set(0)
    
    # # alternative nonstructural rank 2 x
    # key3, key4 = jr.split(key1)
    # x = jr.normal(key1, (5,2)) @ jr.normal(key1, (2,3))

def run(x, y):
    sol = lx.linear_solve(lx.MatrixLinearOperator(x), y, lx.QR(), throw=False)
    return sol.value, sol.result

def run2(x, y):
    sol = lx.linear_solve(lx.MatrixLinearOperator(x), y, lx.SVD(), throw=False)
    return sol.value, sol.result

out_qr, result_qr = run(x, y)
out_svd, result_svd, = run2(x, y)
assert result_svd == lx.RESULTS.successful

rank = jnp.linalg.matrix_rank(x)
is_full_rank = rank == min(x.shape)
assert is_full_rank == want_full_rank

print(f"Got matrix {x} with rank {rank} which is{'' if is_full_rank else ' not'} full rank.")
if result_qr == lx.RESULTS.successful:
    assert jnp.allclose(out_qr, out_svd)
    print("QR solved succeeded.")
else:
    print("QR solve failed.")
```

[the remark this answers is not in the document]
Comment (adconner) [reply lineax/_solve.py]: The contract by jax/numpy I imagine mirrors the same contract as in the underlying LAPACK routines, eg [here](https://www.netlib.org/lapack/explore-html/d8/d83/group__gels_ga73287c9230a903692bc7c933af0da282.html)
```
!> It is assumed that A has full rank, and only a rudimentary protection
!> against rank-deficient matrices is provided. This subroutine only detects
!> exact rank-deficiency, where a diagonal element of the triangular factor
!> of A is exactly zero.
```
Of course, even in the exact rank deficient case like the structural rank 2 above, floating point roundoff error during the calculation can make these entries small nonzero. So the impression one gets is there are two kinds of solvers, ones assuming full rank input which are allowed to divide by values very close to zero assuming they really are meant to be nonzero, and ones robust to non-full rank inputs (rank revealing algorithms).

My suggestions are documented in our discussion in #158: I would be in favor of a assume_full_rank tag for operators which the first kind of solver would filter for at compile time and the second kind of solver would verify at runtime and otherwise error out. As I mention in our discussion the cases where generality are lost are both necessarily quite trivial and also easily recovered by factoring the matrix appropriately. If you wish, you can also introduce operator tags assume_constant_row_space and assume_constant_column_space which control the corresponding jvp terms (and are appropriately implied by assume_full_rank and shape). You could even make a method to guess these tags for a given parameterized operator.

I'm also open to assume_full_rank being a property of the solvers like in #158, but this structure seems slightly logically less pleasing.

But also you can just leave it as is if backward compatibility is a deciding factor. In this case allow_dependent_{rows,columns} means nothing when True and when False, they mean "assume constant row/column space". Every operator logically does a pseudoinverse solve and has undefined behavior if its declared assumptions aren't met.

**Comment by patrick-kidger** [comment]
Closing now that #158 is in, we'll merge #159 instead.
